gap/attributes/dual.gi

```diff
@@ -0,0 +1,213 @@
+#############################################################################
+##
+##  dual.gi
+##  Copyright (C) 2018                                      James D. Mitchell
+##                                                          Finn Smith
+##
+##  Licensing information can be found in the README file of this package.
+##
+#############################################################################
+##
+## This file contains an implementation of dual semigroups.
+## We only provide enough functionality to allow dual semigroups to work
+## as enumerable semigroups.
+## This is to avoid having to install versions of every function in Semigroups
+## specially for dual semigroup representations.
+## In some cases special functions would be faster.
+
+InstallMethod(DualSemigroup, "for a semigroup",
+[IsSemigroup],
+function(S)
+  local dual, fam, filts, type;
+
+  if IsDualSemigroupRep(S) then
+    if HasGeneratorsOfSemigroup(S) then
+      return Semigroup(List(GeneratorsOfSemigroup(S),
+                            x -> UnderlyingElementOfDualSemigroupElement(x)));
+    fi;
+    ErrorNoReturn("Semigroups: DualSemigroup: \n",
+                  "this dual semigroup cannot be constructed ",
+                  "without knowing generators,");
+  fi;
+
+  fam   := NewFamily("DualSemigroupElementsFamily", IsDualSemigroupElement);
+  dual  := Objectify(NewType(CollectionsFamily(fam),
+                            IsWholeFamily and
+                            IsDualSemigroupRep and
+                            IsAttributeStoringRep),
+                    rec());
+
+  filts := IsDualSemigroupElement;
+  if IsMultiplicativeElementWithOne(Representative(S)) then
+    filts := filts and IsMultiplicativeElementWithOne;
+  fi;
+
+  type       := NewType(fam, filts);
+  fam!.type  := type;
+
+  SetDualSemigroupOfFamily(fam, dual);
+
+  SetElementsFamily(FamilyObj(dual), fam);
+  SetDualSemigroup(dual, S);
+
+  if HasIsFinite(S) then
+    SetIsFinite(dual, IsFinite(S));
+  fi;
+
+  if IsTransformationSemigroup(S) then
+    SetAntiIsomorphismTransformationSemigroup(dual,
+                                            AntiIsomorphismDualSemigroup(dual));
```

Comment:
Indentation? If the resulting line is too long, then assign `map := AntiIsomorphismDualSemigroup(dual);` in the previous line